localstack/services/sns/publisher.py

```diff
@@ -0,0 +1,950 @@
+import ast
+import base64
+import datetime
+import json
+import logging
+import time
+import traceback
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Union
+
+import requests
+
+from localstack import config
+from localstack.aws.api.lambda_ import InvocationType
+from localstack.aws.api.sns import MessageAttributeMap
+from localstack.aws.api.sqs import MessageBodyAttributeMap
+from localstack.config import external_service_url
+from localstack.services.sns import constants as sns_constants
+from localstack.services.sns.models import (
+    SnsApplicationPlatforms,
+    SnsMessage,
+    SnsStore,
+    SnsSubscription,
+)
+from localstack.utils.aws import aws_stack
+from localstack.utils.aws.arns import (
+    extract_region_from_arn,
+    extract_resource_from_arn,
+    parse_arn,
+    sqs_queue_url_for_arn,
+)
+from localstack.utils.aws.aws_responses import create_sqs_system_attributes
+from localstack.utils.aws.dead_letter_queue import sns_error_to_dead_letter_queue
+from localstack.utils.cloudwatch.cloudwatch_util import store_cloudwatch_logs
+from localstack.utils.json import json_safe
+from localstack.utils.objects import not_none_or
+from localstack.utils.strings import long_uid, md5, to_bytes
+from localstack.utils.time import timestamp_millis
+
+LOG = logging.getLogger(__name__)
+
+# future config flag
+PLATFORM_APPLICATION_REAL = False
+
+
+@dataclass(frozen=True)
+class SnsPublishContext:
+    message: SnsMessage
+    store: SnsStore
+    request_headers: Dict[str, str]
+
+
+@dataclass
+class SnsBatchPublishContext:
+    messages: List[SnsMessage]
+    store: SnsStore
+    request_headers: Dict[str, str]
```

Comment:
Is it correct that the batch context is not frozen? It would make sense that both behave the same on a first glance

Reply:
Yes, it's because I filter messages in `publish_batch_to_topic`:
```python
ctx.messages = [
    message
    for message in ctx.messages
    if self._should_publish(ctx.store, message, subscriber)
]
```
Agreed that it's bit weird

Reply:
kk was just making sure :+1: